socorro/cron/jobs/archivescraper.py

```diff
@@ -0,0 +1,351 @@
+        """Scrape the candidates/ directory for beta, release candidate, and final releases"""
+        content = self.download(url_path)
+        version_links = [
+            link for link in self.get_links(content)
+            if link['text'][0].isdigit()
+        ]
+
+        # For each version in the candidates/ directory, we traverse the tree finding
+        # the first build info file we can find
+        for link in version_links:
+            content = self.download(link['path'])
+            build_links = [
+                link for link in self.get_links(content)
+                if link['text'].startswith('build')
+            ]
+
+            #  /pub/PRODUCT/candidates/VERSION/...   # noqa
+            # 0/1  / 2     /3         /4
+            version_root = link['path'].split('/')[4]
+            version_root = version_root.replace('-candidates', '')
+
+            for i, build_link in enumerate(build_links):
+                # Get the json file somewhere in this part of the tree
+                json_link = self.get_json_link(build_link['path'])
+                if not json_link:
+                    self.config.logger.warning(
+                        'could not find json file in: %s', build_link['path']
+                    )
+                    continue
+
+                json_file = self.download(json_link)
+                data = json.loads(json_file)
+
+                if 'buildhub' in json_link:
+                    # We have a buildhub.json file to use, so we use that
+                    # structure
+                    data = {
+                        'product_name': product_name,
+                        'release_channel': data['target']['channel'],
+                        'release_version': data['target']['version'],
+                        'build_id': data['build']['id'],
+                        'archive_url': urljoin(self.config.base_url, json_link)
+                    }
+
+                else:
+                    # We have the older build info file format, so we use that
+                    # structure
+                    data = {
+                        'product_name': product_name,
+                        'release_channel': data['moz_update_channel'],
+                        'release_version': data['moz_app_version'],
+                        'build_id': data['buildid'],
+                        'archive_url': urljoin(self.config.base_url, json_link)
+                    }
+
+                # The build link text is something like "build1/" and we
+                # want just the number part, so we drop "build" and the "/"
+                rc_version_string = version_root + 'rc' + build_link['text'][5:-1]
+
+                # Whether or not this is the final build for a set of builds; for
+                # example for [build1, build2, build3] the final build is build3
+                final_build = (i + 1 == len(build_links))
+
+                if final_build:
```

Comment:
This is some mildly crazy logic because of the way release candidates work. archivescraper generates data that's explicitly oriented for the `BetaVersionRule` to do a lookup with a single db query. As such, the release candidates for a final build are in the "release" channel on Buildhub, but in the "beta" channel in `crashstats_productversion` because the crash reports for those builds come in with `ReleaseChannel=beta`.

I don't know if that's a big deal, but it's possible that it makes this data bad for other people to use.